[Core] Add MFU tracking to GPU model execution

bwasti opened the pull request:
This diff calculates MFU in two ways
- Simple: user specified active parameters, using 2ND (2 x num params x input length)
- Advanced, automatic: traversing the compilation graph.  This mode also accumulates information useful for roofline analysis (such as bytes read/written).

Usage:

- `--mfu-analysis-interval`: The interval (in number of steps) between MFU analysis (determining the FLOPs).  Defaults to -1 and never runs.
- `--mfu-analysis-mode`: defaults to "fast".  "detailed" is slower, but more accurate.  "manual" requires user input.
- `--mfu-analysis-active-parameters`: The number of parameters active to run the model (without the un-embedding layer).  This argument applies *only* if `--mfu-analysis-mode` is "manual".  If not specified, vLLM will attempt to derive the active parameter count automatically.

## Purpose

The purpose of this PR is get MFU serving two separate goals:

1. "good enough" MFU -- frequent, highly available MFU data that has zero impact on efficiency.  This MFU data should be extremely cheap to acquire and possibly always on (in the future).

2. "accurate" MFU -- this path optimizes for correctness to ensure we have zero blind-spots.  As model architectures and hardware platforms change we should keep a close eye on the exact implications that has on execution.  This data is expensive to acquire and will likely be used in a more of a performance debugging setup. 

However, these should eventually be the *same*.  The current issue is that it is hard to derive a generic flop counter based on a model, but it should be possible.  The difficulty lies in reliably determining the exact impact inputs have on each matrix multiplication and attention implementation in the model.

For example, given an operation: "mm(A, B)", which argument is a fixed size? which argument is linearly related to the input size? which is quadratic (in the case of attention)?  All of this information *does* exist in the PyTorch graph, but is not in scope for this PR.

## Test Plan

```
pytest -s -v tests/v1/metrics/test_mfu_metrics.py
```

## Test Result

All passed.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

yeqcharlotte reviewed (commented):
great! can we cross validated the difference compared to tensor core utilization? gb/s should be easier to validate with nvidia-smi or dyno.

bwasti commented on vllm/config/__init__.py:
```diff
@@ -383,6 +383,20 @@ class ModelConfig:
     `prompt_embeds` key. Note that enabling this will double the time required
     for graph compilation."""
     served_model_name: Optional[Union[str, list[str]]] = None
+    """Enable periodic MFU calculations every 'mfu_analysis_interval' model 
+    executions.  Defaults to -1, which is disabling the calculation.
+    Setting to 0 does the analysis for every step. """
+    mfu_analysis_interval: int = -1
+    """Setting `mfu_analysis_detailed` to True will capture the torch FX
+    graph and attempt to properly count every flop and byte moved.  This is a
+    very slow operation and should be run infrequently for advanced
+    optimization usecases.  Otherwise, MFU will be flop-only using the 2*N*D
+    approximation (default).
+    N = number of parameters, D = number of new tokens."""
+    mfu_analysis_detailed: bool = False
+    """Used when `mfu_analysis_detailed == False` to specify how to calculate
+    the total flops.  This currently can't be done automatically."""
+    mfu_analysis_active_parameters: float = 0
```
@hmellor +1 will move to observability.

@yeqcharlotte I can swap "detailed" to "mode", but we still need to parameterize the interval + the manual version, right?

bwasti commented:
@yeqcharlotte I've avoided adding the debug for now, since it'd be yet another configuration and I want to keep this clean for the time being.

In a followup PR I plan to dump the MFU per module (by using profile traces), but I haven't figured out the best way to thread that code together.  We luckily have stack traces etc so it's certainly possible!

yeqcharlotte commented on vllm/config/__init__.py:
```diff
@@ -84,6 +84,176 @@ def metrics_info(self) -> dict[str, str]:
+    """Enable periodic MFU calculations every 'mfu_analysis_interval' model 
+    executions.  Defaults to -1, which is disabling the calculation.
+    Setting to 0 does the analysis for every step. """
+    mfu_analysis_active_parameters: float = 0
```
can we add proper validation for arg here?

yeqcharlotte commented on vllm/v1/engine/llm_engine.py:
```diff
@@ -265,7 +265,8 @@ def step(self) -> Union[list[RequestOutput], list[PoolingRequestOutput]]:
 
         # 2) Process EngineCoreOutputs.
-        iteration_stats = IterationStats() if self.log_stats else None
+        iteration_stats = IterationStats(
```
wanna know if @markmc has strong opinion here to do it differently. would love your help to hoop this up with prometheus too!

yeqcharlotte reviewed (commented):
good stuff! i think we should get them in and keep iterating for more coverage.

could also add some docs in https://docs.vllm.ai/en/latest/contributing/profiling.html#example-usage-decorator. and list about model coverage and limitations.

markmc commented on vllm/v1/core/mfu_utils.py:
```diff
@@ -0,0 +1,408 @@
+    return mfu_info
+
+
+class MFUAnalysisLogging:
```
Makes sense to me to put this class under `vllm/v1/metrics`, and any code that is exclusively used by this class

markmc commented on vllm/v1/core/mfu_utils.py:
```diff
@@ -0,0 +1,408 @@
+class MFUAnalysisLogging:
+    """Aggregate and log MFU metrics.
+
+    LoggingStatLogger aggregates per-iteration metrics over a set
```
Typo from copy and paste

markmc commented on vllm/config/__init__.py:
```diff
@@ -84,6 +84,176 @@ def metrics_info(self) -> dict[str, str]:
+            self.collect_detailed_traces[0].split(","))
+
+    mfu_analysis_mode: str = "fast"
+    """Setting `mfu_analysis_mode` to
```
Please expand the MFU acronym somewhere for users - in these help docs would be a good start

markmc commented on vllm/v1/engine/__init__.py:
```diff
@@ -161,6 +173,7 @@ class EngineCoreOutputs(
 
     utility_output: Optional[UtilityOutput] = None
     finished_requests: Optional[set[str]] = None
+    mfu_output: Optional[MFUOutput] = None
```
I think it's probably fine to include this in `SchedulerStats`?

`SchedulerStats` is probably poorly named. AFAIR it's really "not-per-request stats" at this stage

markmc commented on vllm/v1/engine/async_llm.py:
```diff
@@ -439,7 +440,17 @@ async def output_handler():
+                    else:
+                        mfu_info = None
+
+                    iteration_stats = IterationStats(mfu_info=mfu_info) if (
```
Yeah, the purpose of `IterationStats` is to aggregate data as `process_outputs()` iterates across the per-request `EngineCoreOutput` list

None of that aggregation is needed here - if the MFU stats was in `SchedulerStats` you can just use it from there in the logger's `record()` method

markmc commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -994,8 +995,17 @@ def update_from_output(
         # Create EngineCoreOutputs for all clients that have requests with
         # outputs in this step.
+        mfu_output = None
+        if mfu_info:
```
Just pass `mfu_info` to `make_scheduler_stats()` below

markmc reviewed (changes requested):
Thanks for including me. Just some comments from a quick pass of the code

Main comment on the metrics side is that I think you should be able to use `SchedulerStats` for this rather than `IterationStats`, which I admit isn't obvious from the naming

markmc commented:
> ## Code Review
> This pull request introduces MFU (Model Floating-point Unit) tracking for GPU model execution [...]

I'm glad it's not just me for which MFU was a non-obvious acronym! Please expand it liberally :+1:

ProExpertProg commented:
How does this deal with python-defined torch custom ops that obscure the internals of e.g. `fused_moe` and `unified_attention`?

bwasti committed:
Add MFU (Model FLOPs Utilization) analysis feature

This commit adds comprehensive MFU analysis capabilities to vLLM, enabling
users to track and analyze model performance metrics during inference.

Key changes:
- Add MFU analysis utilities in v1/core/mfu_utils.py with detailed and fast modes
- Add MFUAnalyzerMixin for GPU model runner to wrap model execution with MFU tracking
- Add MFU-related configuration options (mfu_analysis_mode, mfu_analysis_interval,
  mfu_analysis_active_parameters) to ObservabilityConfig
- Integrate MFU info throughout the execution pipeline from model runner to
  scheduler to engine outputs
- Add MFU logging and metrics collection in stat loggers
- Add comprehensive tests for MFU analysis with multiple configurations

The MFU analysis supports three modes:
- "fast": Uses 2AND approximation with automatic parameter derivation
- "manual": Uses 2AND with user-specified active parameters
- "detailed": Captures torch FX graph for precise FLOP and byte counting

Signed-off-by: Bram Wasti <bwasti@meta.com>

bwasti committed:
Fix MFU test: handle missing supports_v1 and enable offline mode

- Add check for supports_v1 being None before calling it
- Set HF_HUB_OFFLINE and TRANSFORMERS_OFFLINE env vars for offline mode
- Prevents TypeError when supports_v1 attribute is missing from platform

Signed-off-by: Bram Wasti <bwasti@meta.com>

bwasti committed:
Remove offline mode env vars from MFU test

Environment variables like HF_HUB_OFFLINE, TRANSFORMERS_OFFLINE, and
HF_HUB_DISABLE_XET must be set before the Python process starts,
not inside the test file. Users should set these when running pytest:

  HF_HUB_DISABLE_XET=1 pytest tests/v1/metrics/test_mfu_metrics.py

Signed-off-by: Bram Wasti <bwasti@meta.com>

bwasti committed:
Fix MFU test SimpleStatsLogger signature

Add missing mm_cache_stats parameter to SimpleStatsLogger.record() to match
the StatLoggerBase interface signature.

Signed-off-by: Bram Wasti <bwasti@meta.com>

bwasti committed:
Add error handling for MFU detailed mode and skip incompatible test

- Add try-catch in analyze_model_mfu to handle torch.export failures
- Skip detailed mode test due to torch.export compatibility issues
- Detailed mode will gracefully fall back with warning message
- Manual and fast modes work correctly and all tests pass

Signed-off-by: Bram Wasti <bwasti@meta.com>

hmellor commented on vllm/config/observability.py:
```diff
@@ -114,6 +114,24 @@ def _validate_collect_detailed_traces(
             value = cast(list[DetailedTraceModules], value[0].split(","))
         return value
 
+    mfu_analysis_mode: str = "fast"
```
```py
MFUAnalysisMode = Literal["detailed", "fast", "manual"]
```

```suggestion
    mfu_analysis_mode: MFUAnalysisMode = "fast"
```

hmellor commented on vllm/config/observability.py:
```diff
@@ -114,6 +114,24 @@ def _validate_collect_detailed_traces(
+         automatically deriving active parameter count.
+       - "manual" will require user specified `mfu_analysis_active_parameters`
+         and also uses 2AND to approximate MFU.
+         N = number of parameters, D = number of new tokens."""
```
Necessary for proper rendering in `--help`

```suggestion
    """Setting `mfu_analysis_mode` to"

       - "detailed" will capture the torch FX graph and attempt to properly
         count every flop and byte moved.  This is a very slow operation and
         should be run infrequently for advanced optimization usecases.\n
       - "fast" (default) will attempt to use the 2AND approximation by
         automatically deriving active parameter count.\n
       - "manual" will require user specified `mfu_analysis_active_parameters`
         and also uses 2AND to approximate MFU.
         N = number of parameters, D = number of new tokens."""
```
